### DISCUSSION TITLE
Explorer blocked by the authentication module

### DISCUSSION THREAD

**Discussion by frbuceta** [loopbackio/loopback-next#7166] [Ideas]
## Steps to reproduce

The explorer seems to require authentication when it is set:
```
// Customize @loopback/authentication configuration here
this.configure(AuthenticationBindings.COMPONENT).to({
  defaultMetadata: {
    strategy: 'JWTStrategy'
  }
})
this.component(AuthenticationComponent);
registerAuthenticationStrategy(this, JWTAuthenticationStrategy)
```
![Screenshot 2020-06-17 at 12 20 12](https://user-images.githubusercontent.com/13822438/84887005-73583e00-b095-11ea-8e44-993f7cee289d.png)

## Expected Behavior

Should be set on all controllers, only controllers

## Additional information

```
➜  xxxxxx-monitor-api git:(master) ✗   node -e 'console.log(process.platform, process.arch, process.versions.node)'
  npm ls --prod --depth 0 | grep loopbackdarwin x64 14.4.0
➜  xxxxxx-monitor-api git:(master) ✗   npm ls --prod --depth 0 | grep loopback
├── @loopback/authentication@4.2.7
├── @loopback/boot@2.3.3
├── @loopback/context@3.9.0
├── @loopback/core@2.8.0
├── @loopback/openapi-v3@3.4.3
├── @loopback/repository@2.7.0
├── @loopback/rest@5.1.1
├── @loopback/rest-explorer@2.2.4
├── @loopback/service-proxy@2.3.2
├── loopback-connector-mssql@3.8.0
➜  xxxxxx-monitor-api git:(master) ✗ 
```

## Related Issues

https://github.com/strongloop/loopback-next/issues/4782#issuecomment-639400562

_See [Reporting Issues](http://loopback.io/doc/en/contrib/Reporting-issues.html) for more tips on writing good issues_

**Comment by frbuceta** [comment]
If someone confirms that this should be the case when I have time, I will look for the error

**Comment by jannyHou** [comment]
Thanks @frbuceta I tried with [jwt-todo](https://github.com/strongloop/loopback-next/tree/master/examples/todo-jwt) example, confirmed explorer throws 401.

> Should be set on all controllers, only controllers
Hmm, I have a different opinion on the expected behavior. Not only the controller routes goes through the sequence, there are more ways of defining routes, see [declaring rest routes](https://loopback.io/doc/en/lb4/Routes.html#declaring-rest-routes-with-api-specifications).

Instead of adding controller routes to whitelist, we can consider make a blacklist: explorer and spec endpoints like /openapi.json, /openapi.yaml should be in the list. WDYT?

Like
```ts
this.configure(AuthenticationBindings.COMPONENT).to({
      defaultMetadata: {
        strategy: 'JWTStrategy',
        // or a better property name
        exceptions: ['/explorer', '/openapi.json', '/openapi.yaml']
    }
})
```

**Comment by frbuceta** [comment]
> Thanks @frbuceta I tried with [jwt-todo](https://github.com/strongloop/loopback-next/tree/master/examples/todo-jwt) example, confirmed explorer throws 401.
> 
> > Should be set on all controllers, only controllers
> > Hmm, I have a different opinion on the expected behavior. Not only the controller routes goes through the sequence, there are more ways of defining routes, see [declaring rest routes](https://loopback.io/doc/en/lb4/Routes.html#declaring-rest-routes-with-api-specifications).
> 
> Instead of adding controller routes to whitelist, we can consider make a blacklist: explorer and spec endpoints like /openapi.json, /openapi.yaml should be in the list. WDYT?
> 
> Like
> 
> ```ts
> this.configure(AuthenticationBindings.COMPONENT).to({
>       defaultMetadata: {
>         strategy: 'JWTStrategy',
>         // or a better property name
>         exceptions: ['/explorer', '/openapi.json', '/openapi.yaml']
>     }
> })
> ```

I think it's a good idea but with extensions you can make that list very long. For example, the Health extension could be another case of exception

**Comment by raymondfeng** [comment]
The `src/sequence.ts` in `todo-jwt` needs to be updated per https://github.com/strongloop/loopback-next/blob/master/packages/cli/generators/app/templates/src/sequence.ts.ejs. It should call `InvokeMiddleware` action first and return if the `finished` is true.

**Comment by raymondfeng** [comment]
@jannyHou We need to fix all `src/sequence.ts` in all examples.

**Comment by raymondfeng** [comment]
It should be fixed by https://github.com/strongloop/loopback-next/pull/5807

**Comment by jannyHou** [comment]
@raymondfeng This issue is actually a feature I would say, not a bug.

The explorer is not blocked by default, but after **re-configure** the authentication component using

```ts
this.configure(AuthenticationBindings.COMPONENT).to({
  defaultMetadata: {
    // please note this line enables jwt strategy for ALL endpoints, instead of the decorated ones
    strategy: 'JWTStrategy'
  }
})
this.component(AuthenticationComponent);
registerAuthenticationStrategy(this, JWTAuthenticationStrategy)
```

Your PR to update the templates is valid, but this story is not quite relevant. It's about skipping endpoints like explorer when apply a default strategy to entire app.

**Comment by raymondfeng** [comment]
@jannyHou The explorer endpoints are served by `openApiSpec` middleware. It's a not from a controller.

**Comment by jannyHou** [comment]
@raymondfeng IIUC, the config as:

```ts
this.configure(AuthenticationBindings.COMPONENT).to({
  defaultMetadata: {
    // please note this line enables jwt strategy for ALL endpoints, instead of the decorated ones
    strategy: 'JWTStrategy'
  }
})
```

will apply to all endpoints including the ones served by `openApiSpec` middleware, wouldn't it?

**Comment by mrbatista** [comment]
@raymondfeng #5807 Not resolve the issue.
